crates/solvers/src/domain/solver/baseline.rs

```diff
@@ -69,80 +69,99 @@ impl Baseline {
-                                output,
-                                interactions,
-                                gas: route.gas(),
-                            }
-                            .into_solution(auction.gas_price, sell_token, score)?
-                            .with_id(solution::Id(i as u64))
-                            .with_buffers_internalizations(&auction.tokens),
-                        )
+        for (i, order) in auction.orders.into_iter().enumerate() {
+            let sell_token = auction.tokens.reference_price(&order.sell.token);
+            let Some(user_order) = UserOrder::new(&order) else {
+                continue;
+            };
+            let solution = self.requests_for_order(user_order).find_map(|request| {
+                tracing::trace!(order =% order.uid, ?request, "finding route");
+
+                let route = boundary_solver.route(request, self.max_hops)?;
+                let interactions = route
+                    .segments
+                    .iter()
+                    .map(|segment| {
+                        solution::Interaction::Liquidity(solution::LiquidityInteraction {
+                            liquidity: segment.liquidity.clone(),
+                            input: segment.input,
+                            output: segment.output,
+                            // TODO does the baseline solver know about this optimization?
+                            internalize: false,
+                        })
                     })
-            })
-            .collect()
+                    .collect();
+
+                // The baseline solver generates a path with swapping
+                // for exact output token amounts. This leads to
+                // potential rounding errors for buy orders, where we
+                // can buy slightly more than intended. Fix this by
+                // capping the output amount to the order's buy amount
+                // for buy orders.
+                let mut output = route.output();
+                if let order::Side::Buy = order.side {
+                    output.amount = cmp::min(output.amount, order.buy.amount);
+                }
+
+                let score = solution::Score::RiskAdjusted(solution::SuccessProbability(
+                    self.risk
+                        .success_probability(route.gas(), auction.gas_price, 1),
+                ));
+
+                Some(
+                    solution::Single {
+                        order: order.clone(),
+                        input: route.input(),
+                        output,
+                        interactions,
+                        gas: route.gas(),
+                    }
+                    .into_solution(auction.gas_price, sell_token, score)?
+                    .with_id(solution::Id(i as u64))
+                    .with_buffers_internalizations(&auction.tokens),
+                )
+            });
+            if let Some(solution) = solution {
```

Comment:
New code is only from here to the bottom of the function. The rest is copy paste.